File: api/src/main/java/gg/mineral/api/plugin/MineralPlugin.java
package gg.mineral.api.plugin;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import gg.mineral.api.command.Command;
import gg.mineral.api.plugin.listener.Listener;
import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
import lombok.Getter;
import lombok.SneakyThrows;
import lombok.val;

@Getter
public abstract class MineralPlugin {
    private final List<Listener> listeners = new ArrayList<>();
    private final Map<String, Command> commands = new Object2ObjectOpenHashMap<>();

    @SneakyThrows
    public void registerListener(Class<?>... listenerClasses) {
        for (val listenerClass : listenerClasses) {
            val classLoader = listenerClass.getClassLoader();
            val generated = classLoader.loadClass(listenerClass.getName() + "_Generated").asSubclass(Listener.class);
            val listener = generated.getDeclaredConstructor().newInstance();
            this.listeners.add(listener);
        }
    }

    public void registerCommand(Command... commands) {
        for (Command command : commands)
            this.commands.put(command.getName(), command);
    }

    public boolean hasCommand(String name) {
        return commands.containsKey(name);
    }

    public abstract void onEnable();

    public abstract void onDisable();
}

File: api/src/main/java/gg/mineral/api/plugin/MineralPlugin.kt
package gg.mineral.api.plugin

import gg.mineral.api.command.Command
import gg.mineral.api.plugin.listener.Listener
import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap
import lombok.Getter
import lombok.SneakyThrows

@Getter
abstract class MineralPlugin {
    private val listeners: MutableList<Listener> = ArrayList()
    private val commands: MutableMap<String, Command> = Object2ObjectOpenHashMap()

    @SneakyThrows
    fun registerListener(vararg listenerClasses: Class<*>) {
        for (listenerClass in listenerClasses) {
            val classLoader = listenerClass.classLoader
            val generated = classLoader.loadClass(listenerClass.name + "_Generated").asSubclass(
                Listener::class.java
            )
            val listener = generated.getDeclaredConstructor().newInstance()
            listeners.add(listener)
        }
    }

    fun registerCommand(vararg commands: Command) {
        for (command in commands) this.commands[command.getName()] =
            command
    }

    fun hasCommand(name: String): Boolean {
        return commands.containsKey(name)
    }

    abstract fun onEnable()

    abstract fun onDisable()
}
